engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/ThrowPanicNode.java

```diff
@@ -1,16 +1,74 @@
 package org.enso.interpreter.node.expression.builtin.error;
 
+import com.oracle.truffle.api.dsl.Fallback;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.library.CachedLibrary;
 import com.oracle.truffle.api.nodes.Node;
 import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.builtin.Builtins;
+import org.enso.interpreter.runtime.callable.atom.Atom;
 import org.enso.interpreter.runtime.error.PanicException;
 import org.enso.interpreter.runtime.state.Stateful;
 
 @BuiltinMethod(
     type = "Panic",
     name = "throw",
     description = "Throws a new Panic with given payload.")
-public class ThrowPanicNode extends Node {
-  Stateful execute(Object _this, Object payload) {
-    throw new PanicException(payload, this);
+public abstract class ThrowPanicNode extends Node {
+  static ThrowPanicNode build() {
+    return ThrowPanicNodeGen.create();
+  }
+
+  abstract Stateful execute(Object _this, Object payload);
+
+  @Specialization
+  Stateful doAtom(
+      Object _this, Atom payload, @CachedLibrary(limit = "5") InteropLibrary interopLibrary) {
+    Builtins builtins = Context.get(this).getBuiltins();
+    if (payload.getConstructor() == builtins.caughtPanic()) {
+      // Note [Original Exception Type]
+      Object originalException = payload.getFields()[1];
+      if (interopLibrary.isException(originalException)) {
+        try {
+          throw interopLibrary.throwException(originalException);
+        } catch (UnsupportedMessageException e) {
+          throw new IllegalStateException(
+              "Impossible, `isException` returned true for `originalException`.");
+        }
+      } else {
+        throw new PanicException(
```

Comment:
It should never happen anyway - because `throwException` says it can throw UnsupportedMessage only if `isException` was false and we check it to be true just a moment earlier. But I changed it still as you suggested to throw a panic.